src/geometry/manifold/manifold.lean

```diff
@@ -0,0 +1,539 @@
+  comp     := λe e' he he', begin
+    split,
+    { apply PG.comp he.1 he'.1,
+      apply e.continuous_to_fun.preimage_open_of_open e.open_source e'.open_source },
+    { apply PG.comp he'.2 he.2,
+      apply e'.continuous_inv_fun.preimage_open_of_open e'.open_target e.open_target }
+  end,
+  inv      := λe he, ⟨he.2, he.1⟩,
+  id_mem   := ⟨PG.id_mem, PG.id_mem⟩,
+  locality := λe he, begin
+    split,
+    { apply PG.locality e.open_source (λx xu, _),
+      rcases he x xu with ⟨s, s_open, xs, hs⟩,
+      refine ⟨s, s_open, xs, _⟩,
+      convert hs.1,
+      exact (interior_eq_of_open s_open).symm },
+    { apply PG.locality e.open_target (λx xu, _),
+      rcases he (e.inv_fun x) (e.map_target xu) with ⟨s, s_open, xs, hs⟩,
+      refine ⟨e.target ∩ e.inv_fun ⁻¹' s, _, ⟨xu, xs⟩, _⟩,
+      { exact continuous_on.preimage_open_of_open e.continuous_inv_fun e.open_target s_open },
+      { rw [← inter_assoc, inter_self],
+        convert hs.2,
+        exact (interior_eq_of_open s_open).symm } },
+  end,
+  eq_on_source := λe e' he ee', begin
+    split,
+    { apply PG.congr e'.open_source ee'.2,
+      have Z := ee'.1,
+      erw ee'.1,
+      exact he.1 },
+    { have A := eq_on_source_symm ee',
+      apply PG.congr e'.symm.open_source A.2,
+      convert he.2,
+      rw A.1,
+      refl }
+  end }
+
+lemma mem_groupoid_of_pregroupoid (PG : pregroupoid H) (e : local_homeomorph H H) :
+  e ∈ groupoid_of_pregroupoid PG ↔ PG.P e.to_fun e.source ∧ PG.P e.inv_fun e.target :=
+by refl
+
+lemma groupoid_of_pregroupoid_le (PG₁ PG₂ : pregroupoid H) (h : ∀f s, PG₁.P f s → PG₂.P f s) :
+  groupoid_of_pregroupoid PG₁ ≤ groupoid_of_pregroupoid PG₂ :=
+begin
+  assume e he,
+  rw mem_groupoid_of_pregroupoid at he ⊢,
+  exact ⟨h _ _ he.1, h _ _ he.2⟩
+end
+
+/-- The groupoid of all local homeomorphisms on a topological space H -/
+def continuous_groupoid (H : Type*) [topological_space H] : structure_groupoid H :=
+groupoid_of_pregroupoid
+{ P        := λf s, true,
+  comp     := λf g u v hf hg huv, trivial,
+  id_mem   := trivial,
+  locality := λf u u_open h, trivial,
+  congr    := λf g u u_open hcongr hf, trivial }
+
+/-- Every structure groupoid is contained in the groupoid of all local homeomorphisms -/
+instance : lattice.order_top (structure_groupoid H) :=
+{ top    := continuous_groupoid H,
+  le_top := λ u f hf, by { split; exact dec_trivial },
+  ..structure_groupoid.partial_order }
+
+end groupoid
+
+section manifold
+
+/-- A manifold is a topological space endowed with an atlas, i.e., a set of local homeomorphisms
+taking value in a model space H, called charts, such that the domains of the charts cover the whole
+space. We express the covering property by chosing for each x a member `chart_at x` of the atlas
+containing x in its source: in the smooth case, this is convenient to construct the tangent bundle
+in an efficient way.
+-/
+class manifold (H : Type*) [topological_space H] (M : Type*) [topological_space M] :=
```

Comment:
We should consider what the spell a user will cast to summon a manifold will look like. At the moment I think it's something like
```
variables (M : Type) [topological_space M] [manifold R^3 M] [has_groupoid R^3 M contact]
```
(imagining `contact : \Pi {n : \nat}, structure_groupoid R^{2n+1}`)

I do like that that both the words `manifold` and `contact` would appear in the spell; changing `manifold` to `charted_space` by itself will make it less clear what is being introduced. But perhaps I'm misunderstanding what the use site would look like still.

Reply:
I don't think it is a technical intermediate object. As I explain in the docstring, to have say a smooth manifold you need the atlas, and then the statement that it is compatible with the smooth groupoid. Same for topological manifold, contact manifold, and so on. The common denominator to all these expressions is the word "manifold" and the common denominator to these definitions is the atlas. So, for me, a manifold is exactly a space with an atlas. No more, no less.

Reply:
At some point you'll need to introduce the actual definition of a smooth manifold for instance. And it can't be the spell written by Scott. How will maximality (or equivalence classes) will enter the game? We don't want that adding a redundant chart changes the manifold structure in Lean, do we?

Reply:
It will be (a version of) the spell by Scott: with the current design, the atlas will not be changed, and it will not be maximal. One could then define `maximal_atlas M G` where `G` is a structure groupoid on the model space `H`, as all the charts that are compatible with the atlas and the groupoid `G`, and formulate statements in terms of the maximal atlas without touching the original one. It turns out that, contrary to what I expected at the start, I have never needed the maximal atlas until now.

Reply:
So you intend to have a funny equality defined on manifolds, analogous to the equivalence relation you put on local homeomorphisms?

Reply:
It's already there, and it is called diffeomorphism (well, structomorphism). Note that the definition of a structomorphism is *not* that it induces a bijection between the atlases, but really that read in the charts it belongs to the groupoid. One could show that it is equivalent to the fact that it induces a bijection between the maximal atlases.

Reply:
No, the relation I'm talking about is equivalent to *identity* being a structomorphism.

Reply:
Yes, it is exactly that. Although in general, I am not expecting to put several manifold structures on the same space (I would rather rename the space, probably). (In the same way that we usually don't put on the same space several topological structures that are not defeq)

Reply:
Then it is not really equality on manifolds, but instead equality on manifold structures on a fixed type (or set). Does this concept come up much in practice?

I was also concerned about this point when I looked at this, but I think it's not that big a deal. It's true there are "more distinct manifolds" according to `=` than there would be with the maximal atlas definition, but then again replacing the underlying type of a manifold with an isomorphic type also results in an unequal manifold, and that would be true under either definition, so if you like, you could consider the set-theoretic interpretation of a `manifold` to encode the atlas structure in the choice of underlying set.

However, one thing to watch out for is that, since structomorphism is not the same as structural isomorphism (i.e., an isomorphism preserving the atlases exactly), it is easy to make definitions which are not structomorphism-invariant. For example, `basic_smooth_bundle_core` is not structomorphism-invariant. If we have a smooth bundle on a manifold and then we add a chart over which the bundle cannot be trivialized, there is no corresponding bundle on the new manifold.

Reply:
This is all pretty worrying, but I'm ready to believe I should wait to see more of the theory. In this mean time we could add big warning signs but merge anyway.

Reply:
Indeed, with these definitions there is some work involved to show, say, that a diffeomorphism induces a diffeomorphism of tangent bundles, as a diffeomorphism does not transport atlas to atlas. This means that, even if we had a perfect transfer tactic, there would still be some work here. For this specific example, this just comes from general API on derivatives, so in fact there is no pain.

Reply:
Could you write documentation somehow recording this discussion?

Reply:
Done